src/python/WMCore/MicroService/Unified/MSOutputTemplate.py

```diff
@@ -0,0 +1,228 @@
+import time
+
+from copy import deepcopy
+from bson import Timestamp, ObjectId
+
+
+class MSOutputTemplate(dict):
+    """
+    A simple template to represent the objects stored in Mongodb for MSOutput
+    Its purpose should be to serve as a bidirectional representation of the
+    objects stored in Mongodb. The instances of this class should be used as a
+    buffer for both inbound (reading from the database) and outbound (writing to
+    database) documents. And it must have all the methods needed to set or update
+    various fields in the object. So that we assure the uniformity of all documents.
+
+    Template format:
+    {
+    "RequestName": "ReqName",
+    "RequestStatus": "(clompletd|cloed-out|announced),
+    "creationTime": integer timestamp,
+    "lastUpdate": integer timestamp,
+    "isRelVal": (True|False),
+    "isTaken": (True|False),
+    "destination": ["list of locations"],
+    "OutputDatasets": ["list of datasets"],
+    "destinationOutputMap": [{"destination": ["list of locations"],
+                              "datasets": ["list of datasets"]},
+                             {"destination": ["list of locations"],
+                              "datasets": ["list of datasets"]}],
+    "campaignOutputMap": [{"campaignName": "blah",
+                           "datasets": ["list of datasets"]},
+                          {"campaignName": "blah",
+                           "datasets": ["list of datasets"]}],
+    "transferStatus": "blah", # either "pending" or "done",
+    "transferIDs": ["list of transfer IDs"],
+    "numberOfCopies": integer
+    }
+    """
+    # TODO:
+    #    To add an identifier parameter to __init__
+    #    to distinguish inbound from outbound documents.
+
+    # DONE:
+    #    to figure out how to pass params as dict not only kwargs so I can use
+    #    the same logic for both inbound and outbound documents
+
+    # TODO:
+    #    to implement document mutex in a model similar to the one described here:
+    #    https://stackoverflow.com/questions/32728670/mutex-with-mongodb
+    #    ```
+    #    db.collection.update({done: false, taken: false},{$set: {taken: true, takenBy: myIp}});
+    #    ```
+
+    # TODO:
+    #    as soon as any change in the instantiated object happens ( any call of
+    #    a set method) the same change must be updated in Mongodb too -
+    #    this may happen from outside or from the object itself. The former method
+    #    keeps the document independent from the underlying Database technology
+
+    def __init__(self, doc=None, **kwargs):
+        super(MSOutputTemplate, self).__init__(**kwargs)
+        # super(MSOutputTemplate, self).__init__()
+
+        docTemplate = [
```

Comment:
See my comments from the docstring above. These need to be updated.
BTW, can we change the data type from `unicode` to `str`? Unicode does not exist in the py3 world.

Reply:
Actually, No from Mongo they come as `unicode` - it was before a `str` here, but it was breaking badly because of that so I had to change it to `unicode`. I made a comment in the code regarding this exactly [1]

[1]
https://github.com/dmwm/WMCore/blob/eb2edafab1639a248d1771a1a344e0317710355e/src/python/WMCore/MicroService/Unified/MSOutputTemplate.py#L143-L144